### PR TITLE
Uvs 117: Implement file deletion API

### LINKED ISSUE #117: Implement file deletion API
SP: 3

Server
- [x] file_management -> File deletion
- [x] file_management -> File confirmation
- [x] add playlists to get files/file requests

### PR DESCRIPTION
**Pull request by VolodymyrFihurniak**

### CODE REVIEW

File: server/sql/deleteUserFile.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+DELETE FROM user_files
+WHERE
+  user_id = $1
```
Comment (denbykov) [inline server/sql/deleteUserFile.sql]: Why would you need a user_id here if you already have an id?

File: server/src/business/fileWorker.ts [hunk 2]
```diff
@@ -152,15 +155,59 @@ export class FileWorker {
...
-    await this.db.confirmFile(id, user.id, deviceId);
+    const existFile = await this.db.getUserFile(user.id, fileId);
+    const existSync = await this.db.getSyncrhonizationRecordsByDevice(
+      deviceId,
+      existFile!.id
+    );
+    const existPlaylistFile = await this.playlistDb.getUserPlaylistFile(
+      fileId,
+      user.id,
+      existFile!.id
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: This code does not make any sense

File: server/src/business/fileWorker.ts [hunk 3]
```diff
@@ -152,15 +155,59 @@ export class FileWorker {
...
   };
 
   public confirmFile = async (
-    id: string,
+    fileId: string,
     user: User,
     deviceId: string
   ): Promise<void> => {
-    await this.db.confirmFile(id, user.id, deviceId);
+    const existFile = await this.db.getUserFile(user.id, fileId);
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: Rename to existsingFile

File: server/src/business/fileWorker.ts [hunk 4]
```diff
@@ -152,15 +155,59 @@ export class FileWorker {
...
+    const existSync = await this.db.getSyncrhonizationRecordsByDevice(
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: Rename to existsingSync

File: server/src/business/fileWorker.ts [hunk 5]
```diff
@@ -152,15 +155,59 @@ export class FileWorker {
...
+      deviceId,
+      existFile!.id
+    );
+    const existPlaylistFile = await this.playlistDb.getUserPlaylistFile(
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: You have to fetch all the records here, not a single one.

File: server/src/business/fileWorker.ts [hunk 6]
```diff
@@ -152,15 +155,59 @@ export class FileWorker {
...
+    const userFiles = await this.db.getUserFilesByFileId(existFile!.id);
+    if (userFiles.length) {
+      return;
+    }
+    userFiles.forEach(async (userFile) => {
+      await this.db.deleteUserFile(userFile.userId, userFile.fileId);
+      const file = await this.db.getFile(userFile.fileId);
+      await fs.unlink(file!.path);
+      await this.db.deleteFileById(userFile.fileId);
+    });
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: The rest is not able to make sense as AC 1.3 is compromised

**Review by denbykov — CHANGES_REQUESTED** [review]

File: server/src/business/fileWorker.ts (hunk 2 above)
Comment (VolodymyrFihurniak) [reply server/src/business/fileWorker.ts]: AC 1.3

File: server/src/business/fileWorker.ts [hunk 7]
```diff
@@ -152,15 +154,58 @@ export class FileWorker {
...
+    );
+
+    const existSyncByUserFile =
+      await this.db.getSyncrhonizationRecordsByUserFile(existsingFile!.id);
+
+    if (existSyncByUserFile.isSynchronized) {
+      return;
+    }
+
+    await this.db.deleteUserFile(existsingFile!.id);
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: It should be deleted only for current user

File: server/src/business/fileWorker.ts [hunk 8]
```diff
@@ -152,15 +154,58 @@ export class FileWorker {
...
+    const userFiles = await this.db.getUserFilesByFileId(existsingFile!.id);
+    if (userFiles.length) {
+      return;
+    }
+    userFiles.forEach(async (userFile) => {
+      await this.db.deleteUserFile(userFile.fileId);
+      const file = await this.db.getFile(userFile.fileId);
+      await fs.unlink(file!.path);
+      await this.db.deleteFileById(userFile.fileId);
+    });
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: This loop does not work. Instead you should delete only one file by fileId

**Review by denbykov — CHANGES_REQUESTED** [review]

File: server/src/business/fileWorker.ts [hunk 9]
```diff
@@ -152,15 +154,78 @@ export class FileWorker {
...
+      user.id
+    );
+    if (userPlaylistsByFile.length > 0) {
+      await this.playlistDb.deleteUserPlaylistsFile(
+        fileId,
+        user.id,
+        userPlaylistsByFile
+      );
+    }
+    await this.tagDb.deleteTagMapping(user.id, fileId);
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: Should be deleted for all users

File: server/src/business/fileWorker.ts [hunk 10]
```diff
@@ -152,15 +154,78 @@ export class FileWorker {
...
+    await this.tagDb.deleteTag(fileId);
+    await this.db.deleteFileById(fileId);
+    try {
+      await fs.unlink(file!.path);
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: Unlink should be moved to the data layer

File: server/src/business/fileWorker.ts [hunk 11]
```diff
@@ -152,15 +154,78 @@ export class FileWorker {
...
+    );
+    if (userPlaylistsByFile.length > 0) {
+      await this.playlistDb.deleteUserPlaylistsFile(
+        fileId,
+        user.id,
+        userPlaylistsByFile
+      );
+    }
+    await this.tagDb.deleteTagMapping(user.id, fileId);
+    await this.tagDb.deleteTag(fileId);
```
Comment (denbykov) [inline server/src/business/fileWorker.ts]: You need to delete images from the filesystem

**Review by denbykov — DISMISSED** [review]

**Review by denbykov — APPROVED** [review]

**Pull request merged**